pulser/simulation/simresults.py

```diff
@@ -108,122 +366,136 @@ def get_final_state(self, reduce_to_basis: Optional[str] = None,
+                "dimension > 3.")
+        # Takes care of numerical artefacts in case sum(weights) != 1
+        weights /= sum(weights)
+        return cast(ArrayLike, weights)
+
+    def detection_from_basis_state(self, N_d: int, shot: str,
+                                   spam: dict[str, float]) -> Counter:
+        """Computes the distribution of states detected when detecting `shot`.
+
+        Part of the SPAM implementation : computes measurement errors.
+
+        Args:
+            N_d (int): Number of times state has been detected.
+            shot (str): Binary string of length the number of atoms of the
+            simulation.
+            spam (dict): Dictionnary gathering the SPAM error
+            probabilities.
+        """
+        n_0 = shot.count('0')
+        n_1 = shot.count('1')
+        eps = spam['epsilon']
+        eps_p = spam['epsilon_prime']
+        # Verified
+        prob_1_to_0 = eps_p * (1 - eps) ** n_0 * (1 - eps_p) ** (n_1 - 1)
+        prob_0_to_1 = eps * (1 - eps) ** (n_0 - 1) * (1 - eps_p) ** n_1
+        probs = [int(shot[i]) * prob_1_to_0 + (1 - int(shot[i]))
+                 * prob_0_to_1 for i in range(len(shot))]
+        probs += [1. - sum(probs)]
+        shots = np.random.multinomial(N_d, probs)
+        detected_dict = {shot: shots[-1]}
+
+        for i in range(len(shot)):
+            if shots[i]:
+                detected_dict[shot[:i] + str(1 - int(shot[i])) + shot[i
+                              + 1:]] = shots[i]
+        return Counter(detected_dict)
+
+    def sampling_with_detection_errors(self, spam: dict[str, float],
+                                       t: float,
+                                       N_samples: int = 1000) -> Counter:
+        """Returns the distribution of states really detected.
+
+        Doesn't take state preparation errors into account.
+        Part of the SPAM implementation.
+
+        Args:
+            spam (dict): Dictionnary gathering the SPAM error
+            probabilities.
+            t (float): Time at which to return the samples.
+            N_samples (int): Number of samples.
+        """
+        sampled_state = self.sample_state(t, N_samples)
+        detected_sample_dict: Counter = Counter()
+        for (shot, N_d) in sampled_state.items():
+            dict_state = self.detection_from_basis_state(N_d, shot, spam)
+            detected_sample_dict += dict_state
+
+        return detected_sample_dict
+
+    def plot(self, op: qutip.Qobj, fmt: str = '', label: str = '') -> None:
+        """Plots the expectation value of a given operator op.
+
+        Args:
+            op (qutip.Qobj): Operator whose expectation value is wanted.
+            fmt (str): Curve plot format.
+            label (str): Axis label.
+            error_bars (bool): False here.
+        """
+        plt.plot(self.sim_times, self.expect([op])[0], fmt, label=label)
+        plt.xlabel('Time (µs)')
+        plt.ylabel('Expectation value')
```

Comment:
What if this is defined in `SimulationResults` instead?

Reply:
There's the need for error bars in the `NoisyResults` case that prevents this method from being factorized I think ? The plot methods are different.

Reply:
I'll make a comment on the `NoisyResults` with a suggestion, assuming this is in `SimulationResults`.